src/cockpit/polkit.py

```diff
@@ -0,0 +1,155 @@
+                                   details: Dict[str, str], cookie: str, identities: List[Identity]):
+        logger.debug('BeginAuthentication: action %s, message "%s", icon %s, details %s, cookie %s, identities %r',
+                     action_id, message, icon_name, details, cookie, identities)
+        # only support authentication as ourselves, as we don't yet have the
+        # protocol plumbing and UI to select an admin user
+        my_uid = os.geteuid()
+        for (auth_type, subject) in identities:
+            if auth_type == 'unix-user' and 'uid' in subject and subject['uid'].value == my_uid:
+                logger.debug('Authentication subject %s matches our uid %d', subject, my_uid)
+                break
+        else:
+            logger.warning('Not supporting authentication as any of %s', identities)
+            return
+
+        user_name = pwd.getpwuid(my_uid).pw_name
+        process = await asyncio.create_subprocess_exec(HELPER_PATH, user_name, cookie,
+                                                       stdin=asyncio.subprocess.PIPE,
+                                                       stdout=asyncio.subprocess.PIPE)
+        try:
+            await self._communicate(process)
+        except asyncio.CancelledError:
+            logger.debug('Cancelled authentication')
+            process.terminate()
+        finally:
+            res = await process.wait()
+            logger.debug('helper exited with code %i', res)
+
+    async def _communicate(self, process: asyncio.subprocess.Process):
+        assert process.stdin
+        assert process.stdout
+
+        while process.returncode is None:
+            line = await process.stdout.readline()
+            logger.debug('Read line from helper: %s', line)
+
+            # usually: PAM_PROMPT_ECHO_OFF Password: \n
+            if line.startswith(b'PAM_PROMPT'):
+                prompt = line.split(b' ', 1)[1].strip().decode()
+                # Don't pass this to the UI if it's "Password" (the usual case),
+                # so that superuser.py uses the translated default
+                if prompt.startswith('Password'):
+                    prompt = ''
+
+                passwd = await self.responder.do_askpass('', prompt, '')
+                if passwd is None:
+                    logger.debug('got PAM_PROMPT %s, but do_askpass returned None', prompt)
+                    raise asyncio.CancelledError('no password given')
+                logger.debug('got PAM_PROMPT %s, do_askpass returned a password', prompt)
+                process.stdin.write(passwd.encode())
+                process.stdin.write(b'\n')
+                del passwd  # don't keep this around longer than necessary
+                await process.stdin.drain()
+                logger.debug('got PAM_PROMPT, wrote password to helper')
+            elif line.startswith(b'PAM_TEXT_INFO'):
+                logger.info(line.decode())
```

Comment:
I feel like this is exactly the sort of stuff you should be accumulating to pass to the `messages` field of the prompt signal.

Reply:
:see_no_evil: yes, of course! done.